# Identify the similarities and dissimilarities between neighborhoods in Toronto and New York City.
###  Where are the main retail centers? Which neighborhoods would be the best locations?



         Data will be retrieved from Foursquare.  This data will identify the most common venues 
         in both cities.  Data for these cities' neighborhoods and their most common venues, lat and long
         A cluster analysis will be performed to help identify the major retail neighborhoods. 
         The clusters will be inspected to see what the top most common venues are and may suggest
         a neighborhood's that favor retail and restaurants.  
         And the clusters that do not appear to be good candidates for retail/restaurants will be
         examined for other business possibilities.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 4>

1. <a href="#item1">Download datasets for Toronto and New York and merge then into one dataset.</a>

2. <a href="#item2">Explore Neighborhoods in New York and Toronto</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters:  Look for simularities  between the two cities</a>    
</font>
</div>

<h1>1) Get the data, and create dataframes for Toronto and New york.  Merge the two into one 'combined dataset'</h1>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

## 1. Download and Explore Dataset

New York dataset.

In [2]:
!wget -q -O 'newyork_data.html' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


Let's take a quick look at the data.

Notice how all the relevant data is in the *features* key, which is basically a list of the neighborhoods. So, let's define a new variable that includes this data.

In [3]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)
    
# newyork_data is a Dictionary data type

In [4]:
#use the Dictionary 'featues to create a list '
NY_neighborhoods_data = newyork_data['features']

Let's take a look at the first item in this list.

In [5]:
NY_neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

#### Tranform the New York data (a list) into a *pandas* dataframe

The next task is essentially transforming this data of nested Python dictionaries into a *pandas* dataframe. So let's start by creating an empty dataframe.

In [6]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe 
NY_neighborhoods = pd.DataFrame(columns=column_names)
NY_neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [7]:

# remember NY_Neighborhoods_data

for data in NY_neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']  
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = data['geometry']['coordinates'][1]
    neighborhood_lon = data['geometry']['coordinates'][0]
    
    NY_neighborhoods = NY_neighborhoods.append({'Borough': borough,
                                                'Neighborhood': neighborhood_name,
                                                'Latitude': neighborhood_lat,
                                                'Longitude': neighborhood_lon}, ignore_index=True)
NY_neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [8]:
print(NY_neighborhoods.shape)

(306, 4)


## Next, let's create a similar dataframe for Toronto
## Eliminate all rows that have 'Not assigned' value for 'Borough'

In [9]:
df_Toronto = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
df_Toronto = pd.DataFrame(df_Toronto[0])
for column in df_Toronto.columns:
    print(column)

Postcode
Borough
Neighbourhood


In [10]:
print(df_Toronto.shape)
df_Toronto_filtered = df_Toronto[df_Toronto['Borough'] != 'Not assigned']
print(df_Toronto_filtered.shape)

(288, 3)
(211, 3)


In [11]:
print(NY_neighborhoods.shape)

(306, 4)


# Use numpy 'where' to replace 'Neighbourhood' values of 'Not assigned' with the 'Borough' value. 

In [12]:
df_Toronto_filtered['Neighbourhood'] = np.where(df_Toronto_filtered['Neighbourhood'] == 'Not assigned',
                                         df_Toronto_filtered['Borough'],
                                         df_Toronto_filtered['Neighbourhood'])
df_Toronto_final=df_Toronto_filtered

/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [13]:
print('df_Toronto_final',type(df_Toronto_final),df_Toronto_final.shape)

df_Toronto_final <class 'pandas.core.frame.DataFrame'> (211, 3)


## Read geographical coordinates of each postal code

In [14]:
geo_postal_code = pd.read_csv('http://cocl.us/Geospatial_data')
df_toronto_postalCodes = pd.DataFrame(geo_postal_code)
df_toronto_postalCodes.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

##   Merge the two Toronto dataframes such that the 'lat' and 'long and 'neighborhood' data are combined 
## into a new dataframe 'toronto'

In [15]:
toronto = pd.merge(df_Toronto_final,df_toronto_postalCodes,left_on= 'Postcode', right_on='Postal Code')
toronto.head()
print('size toronto {} '.format(toronto.shape))

size toronto (211, 6) 


In [16]:
toronto.head()

Postcode           Borough     Neighbourhood Postal Code   Latitude  \
0      M3A        North York         Parkwoods         M3A  43.753259   
1      M4A        North York  Victoria Village         M4A  43.725882   
2      M5A  Downtown Toronto      Harbourfront         M5A  43.654260   
3      M5A  Downtown Toronto       Regent Park         M5A  43.654260   
4      M6A        North York  Lawrence Heights         M6A  43.718518   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.360636  
4 -79.464763

In [17]:
print('size toronto {} '.format(toronto.shape))

size toronto (211, 6) 


In [18]:
toronto.drop(['Postcode','Postal Code'], 1,inplace = True)

In [19]:
toronto.rename(columns={'Neighbourhood':'Neighborhood'},inplace=True)

## Next, add a column called 'City' to both the New York and Toronto dataframes

In [20]:
toronto['City'] = "Totonto"
NY_neighborhoods['City'] = 'New York City'

## Next let's merge the Toronto and New York dataframe into one master dataframe

In [21]:
df_master = pd.DataFrame(NY_neighborhoods)

In [22]:
df_combined = pd.DataFrame(df_master.append(toronto,ignore_index=True))

In [23]:
df_combined.shape

(517, 5)

## The Toronto and NY dataframes have been combined to create a new DataFrame 'all_data'

In [24]:

all_data = df_combined
all_data.head()


Borough Neighborhood   Latitude  Longitude           City
0   Bronx    Wakefield  40.894705 -73.847201  New York City
1   Bronx   Co-op City  40.874294 -73.829939  New York City
2   Bronx  Eastchester  40.887556 -73.827806  New York City
3   Bronx    Fieldston  40.895437 -73.905643  New York City
4   Bronx    Riverdale  40.890834 -73.912585  New York City

In [25]:
CLIENT_ID = 'HPNNZXX1CG5QXNX20LUFZMXS2S30YSY4RTL1JLRGFWFHXKKE' # your Foursquare ID
CLIENT_SECRET = 'J53PPH0E5AS4YWAE5ANDJUV1WCI3N0PJ5FLPJJ03XMEPIDD3' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: HPNNZXX1CG5QXNX20LUFZMXS2S30YSY4RTL1JLRGFWFHXKKE
CLIENT_SECRET:J53PPH0E5AS4YWAE5ANDJUV1WCI3N0PJ5FLPJJ03XMEPIDD3


##  Get venues.

In [26]:
# a function for retrieving venues 

LIMIT = 500
radius = 300

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)    
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],         
            v['venue']['categories'][0]['name']) for v in results])
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude',
                  'Venue Category']
    
    return(nearby_venues)

<h1>Get venues for the Toronto/New York dataframe</h1>

In [27]:
#
all_venues = getNearbyVenues(names=all_data['Neighborhood'],
                             latitudes=all_data['Latitude'],
                             longitudes=all_data['Longitude'],
                             )

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Marble Hill
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker

In [28]:
print(type(all_venues))
all_venues.head()

<class 'pandas.core.frame.DataFrame'>


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

                                     Venue  Venue Latitude  Venue Longitude  \
0                         Lollipops Gelato       40.894123       -73.845892   
1                                 Rite Aid       40.896649       -73.844846   
2                         Carvel Ice Cream       40.890487       -73.848568   
3  Cooler Runnings Jamaican Restaurant Inc       40.898276       -73.850381   
4                                    Shell       40.894187       -73.845862   

         Venue Category  
0          Dessert Shop  
1              Pharmacy  
2        Ice Cream Shop  
3  Caribbean Restaurant  
4           Gas Station

In [29]:
all_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                               
Adelaide                                                             100   
Agincourt                                                              5   
Agincourt North                                                        3   
Albion Gardens                                                         9   
Alderwood                                                              9   
Allerton                                                              36   
Annadale                                                              11   
Arden Heights                                                          5   
Arlington                                                              5   
Arrochar                                                              19   
Arverne                                                               17   
Astoria                                                              100   
Astoria Heights                                                       13   
Auburndale                                                            20   
Bath Beach                                                            46   
Bathurst Manor                                                        18   
Bathurst Quay                                                         15   
Battery Park City                                                    100   
Bay Ridge                                                             86   
Bay Terrace                                                           53   
Baychester                                                            25   
Bayside                                                               73   
Bayswater                                                              3   
Bayview Village                                                        4   
Beaumond Heights                                                       9   
Bedford Park                                                          61   
Bedford Stuyvesant                                                    27   
Beechhurst                                                            14   
Bellaire                                                              11   
Belle Harbor                                                          18   
Bellerose                                                             21   
Belmont                                                               94   
Bensonhurst                                                           35   
Berczy Park                                                           55   
Bergen Beach                                                           7   
Birch Cliff                                                            4   
Blissville                                                            21   
Bloomfield                                                             5   
Bloordale Gardens                                                      8   
Boerum Hill                                                           87   
Borough Park                                                          19   
Breezy Point                                                           4   
Briarwood                                                             11   
Brighton Beach                                                        43   
Broad Channel                                                          4   
Broadway Junction                                                     14   
Brockton                                                              21   
Bronxdale                                                             14   
Brooklyn Heights                                                     100   
Brookville                                                             1   
Brownsville                                                           19   
Bulls Head             

In [30]:
print('There are {} uniques categories.'.format(len(all_venues['Venue Category'].unique())))

There are 466 uniques categories.


In [31]:
# one hot encoding
all_onehot = pd.get_dummies(all_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
all_onehot['Neighborhood'] = all_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [all_onehot.columns[-1]] + list(all_onehot.columns[:-1])
all_onehot = all_onehot[fixed_columns]
print(all_onehot.shape)
all_onehot.head()

(14803, 466)


Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  \
0            0                  0               0                  0   
1            0                  0               0                  0   
2            0                  0               0                  0   
3            0                  0               0                  0   
4            0                  0               0                  0   

   African Restaurant  Airport  Airport Food Court  Airport Gate  \
0                   0        0                   0             0   
1                   0        0                   0             0   
2                   0        0                   0             0   
3                   0        0                   0             0   
4                   0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  Airport Tram  \
0               0                0                 0             0   
1               0                0                 0             0   
2               0                0                 0             0   
3               0                0                 0             0   
4               0                0                 0             0   

   American Restaurant  Animal Shelter  Antique Shop  Aquarium  Arcade  \
0                    0               0             0         0       0   
1                    0               0             0         0       0   
2                    0               0             0         0       0   
3                    0               0             0         0       0   
4                    0               0             0         0       0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0                 0                       0            0           0   
1                 0                       0            0           0   
2                 0                       0            0           0   
3                 0                       0            0           0   
4                 0                       0            0           0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                    0                     0                 0   
1                    0                     0                 0   
2                    0                     0                 0   
3                    0                     0                 0   
4                    0                     0                 0   

   Athletics & Sports  Auditorium  Australian Restaurant  Austrian Restaurant  \
0                   0           0                      0                    0   
1                   0           0                      0                    0   
2                   0           0                      0                    0   
3                   0           0                      0                    0   
4                   0           0                      0                    0   

   Auto Dealership  Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  \
0                0            0              0                0          0   
1                0            0              0                0          0   
2                0            0              0                0          0   
3                0            0              0                0          0   
4                0            0              0                0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0           0       0     0    0               0   
1           0           0       0     0    0               0   
2           0           0       0     0    0               0   
3           0           0       0     0    0               0   
4           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Bath House  Beach  \
0                 0                 0       

In [32]:
all_grouped = all_onehot.groupby('Neighborhood').mean().reset_index()
print(all_grouped.shape)
all_grouped.head()

(502, 466)


Neighborhood  Yoga Studio  Accessories Store  Adult Boutique  \
0         Adelaide          0.0                0.0             0.0   
1        Agincourt          0.0                0.0             0.0   
2  Agincourt North          0.0                0.0             0.0   
3   Albion Gardens          0.0                0.0             0.0   
4        Alderwood          0.0                0.0             0.0   

   Afghan Restaurant  African Restaurant  Airport  Airport Food Court  \
0                0.0                 0.0      0.0                 0.0   
1                0.0                 0.0      0.0                 0.0   
2                0.0                 0.0      0.0                 0.0   
3                0.0                 0.0      0.0                 0.0   
4                0.0                 0.0      0.0                 0.0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0           0.0             0.0              0.0               0.0   
1           0.0             0.0              0.0               0.0   
2           0.0             0.0              0.0               0.0   
3           0.0             0.0              0.0               0.0   
4           0.0             0.0              0.0               0.0   

   Airport Tram  American Restaurant  Animal Shelter  Antique Shop  Aquarium  \
0           0.0                 0.04             0.0           0.0       0.0   
1           0.0                 0.00             0.0           0.0       0.0   
2           0.0                 0.00             0.0           0.0       0.0   
3           0.0                 0.00             0.0           0.0       0.0   
4           0.0                 0.00             0.0           0.0       0.0   

   Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0     0.0               0.0                     0.0         0.01        0.01   
1     0.0               0.0                     0.0         0.00        0.00   
2     0.0               0.0                     0.0         0.00        0.00   
3     0.0               0.0                     0.0         0.00        0.00   
4     0.0               0.0                     0.0         0.00        0.00   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                  0.0                   0.0              0.02   
1                  0.0                   0.0              0.00   
2                  0.0                   0.0              0.00   
3                  0.0                   0.0              0.00   
4                  0.0                   0.0              0.00   

   Athletics & Sports  Auditorium  Australian Restaurant  Austrian Restaurant  \
0                 0.0         0.0                    0.0                  0.0   
1                 0.0         0.0                    0.0                  0.0   
2                 0.0         0.0                    0.0                  0.0   
3                 0.0         0.0                    0.0                  0.0   
4                 0.0         0.0                    0.0                  0.0   

   Auto Dealership  Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  \
0              0.0          0.0            0.0              0.0        0.0   
1              0.0          0.0            0.0              0.0        0.0   
2              0.0          0.0            0.0              0.0        0.0   
3              0.0          0.0            0.0              0.0        0.0   
4              0.0          0.0            0.0              0.0        0.0   

   Baby Store  Bagel Shop  Bakery  Bank   Bar  Baseball Field  \
0         0.0         0.0    0.03   0.0  0.04             0.0   
1         0.0         0.0    0.00   0.0  0.00             0.0   
2         0.0         0.0    0.00   0.0  0.00             0.0   
3         0.0         0.0    0.00   0.0  0.00             0.0   
4         0.0         0.0    0.00   0.0  0.00             0.0   

   Baseball Stadium

In [33]:
all_grouped.shape

(502, 466)

## Let's look at some some of the top venues for some of the neighborhoods

In [34]:
num_top_venues = 5
cnt = 0
print('len all_grouped',all_grouped.shape)

#print out first 3 neighborhoods
for hood in all_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = all_grouped[all_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')
    cnt += 1
    if cnt == 3:
        break

len all_grouped (502, 466)
----Adelaide----
             venue  freq
0      Coffee Shop  0.06
1             Café  0.05
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Agincourt----
                venue  freq
0              Lounge   0.2
1      Clothing Store   0.2
2      Sandwich Place   0.2
3  Chinese Restaurant   0.2
4      Breakfast Spot   0.2


----Agincourt North----
                   venue  freq
0                   Park  0.67
1             Playground  0.33
2            Yoga Studio  0.00
3  Outdoors & Recreation  0.00
4               Pet Café  0.00




In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

<h1>Top 10 venues for each neighborhood</h1>

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
all_venues_sorted = pd.DataFrame(columns=columns)
all_venues_sorted['Neighborhood'] = all_grouped['Neighborhood']

for ind in np.arange(all_grouped.shape[0]):
    all_venues_sorted.iloc[ind, 1:] = return_most_common_venues(all_grouped.iloc[ind, :], num_top_venues)

all_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0         Adelaide           Coffee Shop                  Café   
1        Agincourt                Lounge        Breakfast Spot   
2  Agincourt North                  Park            Playground   
3   Albion Gardens         Grocery Store           Coffee Shop   
4        Alderwood           Pizza Place           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Bar            Steakhouse       Thai Restaurant   
1    Chinese Restaurant        Clothing Store        Sandwich Place   
2         Women's Store                  Farm     Electronics Store   
3            Beer Store   Fried Chicken Joint           Pizza Place   
4                  Pool                   Gym          Skating Rink   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   American Restaurant                Bakery          Burger Joint   
1         Women's Store    Falafel Restaurant     Electronics Store   
2   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
3              Pharmacy  Fast Food Restaurant        Sandwich Place   
4        Sandwich Place                   Pub              Pharmacy   

  9th Most Common Venue 10th Most Common Venue  
0                 Hotel         Cosmetics Shop  
1   Empanada Restaurant     English Restaurant  
2         Event Service            Event Space  
3               Factory             Eye Doctor  
4           Event Space          Event Service

## 4. Cluster Neighborhoods  - 5 clusters  - Kmeans

In [37]:
# set number of clusters
kclusters = 5

# drop Neighborhood from fequency dataframe so it can used KMeans
all_grouped_clustering = all_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(all_grouped_clustering)
# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 


array([4, 4, 0, 1, 1, 1, 1, 1, 4, 4], dtype=int32)

In [38]:
# add clustering labels to rows of DataFrame
all_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

all_merged = all_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
all_merged = all_merged.join(all_venues_sorted.set_index('Neighborhood'), on='Neighborhood', how='right')

all_merged.head() # check the last columns!

Borough Neighborhood   Latitude  Longitude           City  Cluster Labels  \
0   Bronx    Wakefield  40.894705 -73.847201  New York City               4   
1   Bronx   Co-op City  40.874294 -73.829939  New York City               1   
2   Bronx  Eastchester  40.887556 -73.827806  New York City               4   
3   Bronx    Fieldston  40.895437 -73.905643  New York City               4   
4   Bronx    Riverdale  40.890834 -73.912585  New York City               4   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0          Dessert Shop           Gas Station            Food Truck   
1        Baseball Field           Bus Station  Fast Food Restaurant   
2  Caribbean Restaurant           Bus Station         Metro Station   
3                 Plaza                 River           Bus Station   
4      Greek Restaurant           Coffee Shop        Ice Cream Shop   

  4th Most Common Venue   5th Most Common Venue 6th Most Common Venue  \
0              Pharmacy              Donut Shop        Sandwich Place   
1              Pharmacy                    Park         Grocery Store   
2              Bus Stop                   Diner             Juice Bar   
3            Playground           Women's Store   Egyptian Restaurant   
4    Italian Restaurant  Furniture / Home Store           Bus Station   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Laundromat  Caribbean Restaurant                  Food   
1        Ice Cream Shop        Discount Store          Liquor Store   
2         Bowling Alley    Seafood Restaurant            Donut Shop   
3     Electronics Store   Empanada Restaurant    English Restaurant   
4           Yoga Studio                 Trail             Bookstore   

  10th Most Common Venue  
0         Ice Cream Shop  
1             Restaurant  
2            Pizza Place  
3   Ethiopian Restaurant  
4                   Café

In [39]:
#this is cluster 3
all_merged.loc[all_merged['Cluster Labels'] == 0, all_merged.columns[[1] + list(range(5, all_merged.shape[1]))]]

Neighborhood  Cluster Labels       1st Most Common Venue  \
27           Clason Point               0                        Park   
192            Somerville               0                        Park   
203             Todt Hill               0                        Park   
256         Randall Manor               0                        Park   
303             Bayswater               0                        Park   
306             Parkwoods               0        Fast Food Restaurant   
344   Caledonia-Fairbanks               0                        Park   
366          East Toronto               0                        Park   
376           CFB Toronto               0                     Airport   
377        Downsview East               0                     Airport   
395             Downsview               0  Construction & Landscaping   
396            North Park               0  Construction & Landscaping   
397           Upwood Park               0  Construction & Landscaping   
418         Lawrence Park               0                    Bus Line   
422                Weston               0                        Park   
426       York Mills West               0                        Park   
428     Forest Hill North               0                       Trail   
429      Forest Hill West               0                       Trail   
443     Kingsview Village               0                        Park   
444  Martin Grove Gardens               0                        Park   
445      Richview Gardens               0                        Park   
446          St. Phillips               0                        Park   
461       Agincourt North               0                        Park   
462       L'Amoreaux East               0                        Park   
463              Milliken               0                        Park   
464          Steeles East               0                        Park   
499          The Kingsway               0                  Smoke Shop   
500       Montgomery Road               0                  Smoke Shop   
501        Old Mill North               0                  Smoke Shop   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
27   South American Restaurant         Boat or Ferry              Bus Stop   
192              Women's Store        Farmers Market     Electronics Store   
203              Women's Store        Farmers Market     Electronics Store   
256                   Bus Stop            Bagel Shop           Pizza Place   
303               Tennis Court            Playground         Women's Store   
306          Food & Drink Shop                  Park        Farmers Market   
344              Women's Store              Pharmacy                Market   
366          Convenience Store           Coffee Shop                  Farm   
376                       Park         Women's Store        Farmers Market   
377                       Park         Women's Store        Farmers Market   
395                     Bakery                  Park      Basketball Court   
396                     Bakery                  Park      Basketball Court   
397                     Bakery                  Park      Basketball Court   
418                       Park           Swim School                  Lake   
422          Convenience Store        Farmers Market     Electronics Store   
426          Convenience Store                  Bank        Farmers Market   
428                       Park      Sushi Restaurant         Jewelry Store   
429                       Park      Sushi Restaurant         Jewelry Store   
443          Mobile Phone Shop         Women's Store        Farmers Market   
444          Mobile Phone Shop         Women's Store        Farmers Market   
445          Mobile Phone Shop         Women's Store        Farmers Market   
446          Mobile Phone Shop         Women's Store        Farmers Market   
461                 Playgrou

Finally, let's visualize the resulting clusters

In [40]:
# create map
# I selected 'Binghampton, NY', New York as the center on the map.  
# It is approxiamatelt mid-way between NYC and Toronto 
address = 'Binghampton, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of ''Binghampton, NY'', New York  are {}, {}.'.format(latitude, longitude))
map_clusters = folium.Map(location=[41.0, -74], zoom_start=6)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(all_merged['Latitude'], all_merged['Longitude'], all_merged['Neighborhood'], all_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

The geograpical coordinate of Binghampton, NY, New York  are 42.1147984, -75.8540822.


<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [41]:
cluster_1 = all_merged.loc[all_merged['Cluster Labels'] == 0, all_merged.columns[[1] + list(range(5, all_merged.shape[1]))]]
print(len(cluster_1))
cluster_1

29


Neighborhood  Cluster Labels       1st Most Common Venue  \
27           Clason Point               0                        Park   
192            Somerville               0                        Park   
203             Todt Hill               0                        Park   
256         Randall Manor               0                        Park   
303             Bayswater               0                        Park   
306             Parkwoods               0        Fast Food Restaurant   
344   Caledonia-Fairbanks               0                        Park   
366          East Toronto               0                        Park   
376           CFB Toronto               0                     Airport   
377        Downsview East               0                     Airport   
395             Downsview               0  Construction & Landscaping   
396            North Park               0  Construction & Landscaping   
397           Upwood Park               0  Construction & Landscaping   
418         Lawrence Park               0                    Bus Line   
422                Weston               0                        Park   
426       York Mills West               0                        Park   
428     Forest Hill North               0                       Trail   
429      Forest Hill West               0                       Trail   
443     Kingsview Village               0                        Park   
444  Martin Grove Gardens               0                        Park   
445      Richview Gardens               0                        Park   
446          St. Phillips               0                        Park   
461       Agincourt North               0                        Park   
462       L'Amoreaux East               0                        Park   
463              Milliken               0                        Park   
464          Steeles East               0                        Park   
499          The Kingsway               0                  Smoke Shop   
500       Montgomery Road               0                  Smoke Shop   
501        Old Mill North               0                  Smoke Shop   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
27   South American Restaurant         Boat or Ferry              Bus Stop   
192              Women's Store        Farmers Market     Electronics Store   
203              Women's Store        Farmers Market     Electronics Store   
256                   Bus Stop            Bagel Shop           Pizza Place   
303               Tennis Court            Playground         Women's Store   
306          Food & Drink Shop                  Park        Farmers Market   
344              Women's Store              Pharmacy                Market   
366          Convenience Store           Coffee Shop                  Farm   
376                       Park         Women's Store        Farmers Market   
377                       Park         Women's Store        Farmers Market   
395                     Bakery                  Park      Basketball Court   
396                     Bakery                  Park      Basketball Court   
397                     Bakery                  Park      Basketball Court   
418                       Park           Swim School                  Lake   
422          Convenience Store        Farmers Market     Electronics Store   
426          Convenience Store                  Bank        Farmers Market   
428                       Park      Sushi Restaurant         Jewelry Store   
429                       Park      Sushi Restaurant         Jewelry Store   
443          Mobile Phone Shop         Women's Store        Farmers Market   
444          Mobile Phone Shop         Women's Store        Farmers Market   
445          Mobile Phone Shop         Women's Store        Farmers Market   
446          Mobile Phone Shop         Women's Store        Farmers Market   
461                 Playgrou

#### Cluster 2

In [42]:
cluster_2 = all_merged.loc[all_merged['Cluster Labels'] == 1, all_merged.columns[[1] + list(range(5, all_merged.shape[1]))]]
cluster_2


Neighborhood  Cluster Labels            1st Most Common Venue  \
1             Co-op City               1                   Baseball Field   
8                Norwood               1                      Pizza Place   
14    University Heights               1                      Pizza Place   
15        Morris Heights               1               Spanish Restaurant   
17          East Tremont               1                      Pizza Place   
19          High  Bridge               1                      Pizza Place   
20               Melrose               1                      Pizza Place   
25            Morrisania               1                   Discount Store   
26             Soundview               1                    Grocery Store   
29          Country Club               1                   Sandwich Place   
30           Parkchester               1                      Pizza Place   
32              Van Nest               1                      Pizza Place   
33           Morris Park               1                      Pizza Place   
36       North Riverdale               1                      Pizza Place   
38         Schuylerville               1                         Pharmacy   
40           Castle Hill               1                   Cosmetics Shop   
43             Concourse               1                    Deli / Bodega   
53     Manhattan Terrace               1                      Pizza Place   
55         Crown Heights               1                      Pizza Place   
78          Coney Island               1                      Pizza Place   
80          Borough Park               1                             Bank   
92               Midwood               1                      Pizza Place   
93   Prospect Park South               1             Caribbean Restaurant   
142              Maspeth               1                      Pizza Place   
146            Woodhaven               1                    Deli / Bodega   
147           Ozone Park               1                         Pharmacy   
155            Glen Oaks               1                         Pharmacy   
166             Rochdale               1  Southern / Soul Food Restaurant   
170         Far Rockaway               1                      Pizza Place   
171        Broad Channel               1                  Other Nightlife   
176             Edgemere               1                      Pizza Place   
181          Floral Park               1                Indian Restaurant   
200             Rosebank               1               Italian Restaurant   
205        Port Richmond               1              Rental Car Location   
208    Castleton Corners               1                      Pizza Place   
209      New Springville               1               Chinese Restaurant   
214          Eltingville               1                      Pizza Place   
215             Annadale               1                      Pizza Place   
223                Rugby               1             Caribbean Restaurant   
231         Dongan Hills               1                         Pharmacy   
237      Pleasant Plains               1                      Pizza Place   
240            Rossville               1                      Pizza Place   
241        Arden Heights               1                         Pharmacy   
242           Greenridge               1                       Playground   
243    Heartland Village               1                      Pizza Place   
260             New Lots               1                      Pizza Place   
267    Claremont Village               1                           Bakery   
269           Mount Eden               1                         Pharmacy   
291         Prince's Bay               1                      Pizza Place   
298             Allerton               1                      Pizza Place   
299  Kingsbridge Heights               1                      Pizza Place   
307     

#### Cluster 3

In [43]:
cluster_3 = all_merged.loc[all_merged['Cluster Labels'] == 2, all_merged.columns[[1] + list(range(5, all_merged.shape[1]))]]
cluster_3

Neighborhood  Cluster Labels       1st Most Common Venue  \
255              Emerson Hill               2  Construction & Landscaping   
412                     Emery               2              Baseball Field   
413                 Humberlea               2              Baseball Field   
504                Humber Bay               2  Construction & Landscaping   
505          King's Mill Park               2  Construction & Landscaping   
506  Kingsway Park South East               2  Construction & Landscaping   
507                 Mimico NE               2  Construction & Landscaping   
508            Old Mill South               2  Construction & Landscaping   
509        The Queensway East               2  Construction & Landscaping   
510     Royal York South East               2  Construction & Landscaping   
511                  Sunnylea               2  Construction & Landscaping   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
255        Farmers Market   Egyptian Restaurant     Electronics Store   
412         Women's Store        Farmers Market     Electronics Store   
413         Women's Store        Farmers Market     Electronics Store   
504        Baseball Field  Fast Food Restaurant   Empanada Restaurant   
505        Baseball Field  Fast Food Restaurant   Empanada Restaurant   
506        Baseball Field  Fast Food Restaurant   Empanada Restaurant   
507        Baseball Field  Fast Food Restaurant   Empanada Restaurant   
508        Baseball Field  Fast Food Restaurant   Empanada Restaurant   
509        Baseball Field  Fast Food Restaurant   Empanada Restaurant   
510        Baseball Field  Fast Food Restaurant   Empanada Restaurant   
511        Baseball Field  Fast Food Restaurant   Empanada Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
255   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
412   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
413   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
504    English Restaurant  Ethiopian Restaurant         Event Service   
505    English Restaurant  Ethiopian Restaurant         Event Service   
506    English Restaurant  Ethiopian Restaurant         Event Service   
507    English Restaurant  Ethiopian Restaurant         Event Service   
508    English Restaurant  Ethiopian Restaurant         Event Service   
509    English Restaurant  Ethiopian Restaurant         Event Service   
510    English Restaurant  Ethiopian Restaurant         Event Service   
511    English Restaurant  Ethiopian Restaurant         Event Service   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
255         Event Service           Event Space                Exhibit  
412         Event Service           Event Space                Exhibit  
413         Event Service           Event Space                Exhibit  
504           Event Space               Exhibit             Eye Doctor  
505           Event Space               Exhibit             Eye Doctor  
506           Event Space               Exhibit             Eye Doctor  
507           Event Space               Exhibit             Eye Doctor  
508           Event Space               Exhibit             Eye Doctor  
509           Event Space               Exhibit             Eye Doctor  
510           Event Space               Exhibit             Eye Doctor  
511           Event Space               Exhibit             Eye Doctor

#### Cluster 4

In [44]:
cluster_4 = all_merged.loc[all_merged['Cluster Labels'] == 3, all_merged.columns[[1] + list(range(5, all_merged.shape[1]))]]
cluster_4

Neighborhood  Cluster Labels 1st Most Common Venue  \
322        Cloverdale               3                  Bank   
323         Islington               3                  Bank   
324      Martin Grove               3                  Bank   
325  Princess Gardens               3                  Bank   
326   West Deane Park               3                  Bank   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
322         Women's Store        Farmers Market     Electronics Store   
323         Women's Store        Farmers Market     Electronics Store   
324         Women's Store        Farmers Market     Electronics Store   
325         Women's Store        Farmers Market     Electronics Store   
326         Women's Store        Farmers Market     Electronics Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
322   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
323   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
324   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
325   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
326   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
322         Event Service           Event Space                Exhibit  
323         Event Service           Event Space                Exhibit  
324         Event Service           Event Space                Exhibit  
325         Event Service           Event Space                Exhibit  
326         Event Service           Event Space                Exhibit

#### Cluster 5

In [45]:
cluster_5 = all_merged.loc[all_merged['Cluster Labels'] == 4, all_merged.columns[[1] + list(range(5, all_merged.shape[1]))]]
cluster_5

Neighborhood  Cluster Labels  \
0                                            Wakefield               4   
2                                          Eastchester               4   
3                                            Fieldston               4   
4                                            Riverdale               4   
379                                          Riverdale               4   
5                                          Kingsbridge               4   
6                                          Marble Hill               4   
7                                             Woodlawn               4   
9                                       Williamsbridge               4   
10                                          Baychester               4   
11                                      Pelham Parkway               4   
12                                         City Island               4   
13                                        Bedford Park               4   
406                                       Bedford Park               4   
16                                             Fordham               4   
18                                          West Farms               4   
21                                          Mott Haven               4   
22                                         Port Morris               4   
23                                            Longwood               4   
24                                         Hunts Point               4   
28                                         Throgs Neck               4   
31                                  Westchester Square               4   
34                                             Belmont               4   
35                                      Spuyten Duyvil               4   
37                                          Pelham Bay               4   
39                                      Edgewater Park               4   
41                                           Olinville               4   
42                                      Pelham Gardens               4   
44                                           Unionport               4   
45                                            Edenwald               4   
46                                           Bay Ridge               4   
47                                         Bensonhurst               4   
48                                         Sunset Park               4   
49                                          Greenpoint               4   
50                                           Gravesend               4   
51                                      Brighton Beach               4   
52                                      Sheepshead Bay               4   
54                                            Flatbush               4   
56                                       East Flatbush               4   
57                                          Kensington               4   
58                                     Windsor Terrace               4   
59                                    Prospect Heights               4   
60                                         Brownsville               4   
61                                        Williamsburg               4   
62                                            Bushwick               4   
63                                  Bedford Stuyvesant               4   
64                                    Brooklyn Heights               4   
65                                         Cobble Hill               4   
66                                     Carroll Gardens               4   
67                                            Red Hook               4   
68                                             Gowanus               4   
69                                         Fort Greene               4   
70                                          Park Slope               4   
71                                       Cypre

<h1>Let's add city names to the clusters</h1>

In [46]:
cluster1_cities = pd.merge(cluster_1,df_combined,left_on= 'Neighborhood', right_on='Neighborhood')
cluster2_cities = pd.merge(cluster_2,df_combined,left_on= 'Neighborhood', right_on='Neighborhood')
cluster3_cities = pd.merge(cluster_3,df_combined,left_on= 'Neighborhood', right_on='Neighborhood')
cluster4_cities = pd.merge(cluster_4,df_combined,left_on= 'Neighborhood', right_on='Neighborhood')
cluster5_cities = pd.merge(cluster_5,df_combined,left_on= 'Neighborhood', right_on='Neighborhood')

In [47]:
cluster3_cities.head()

Neighborhood  Cluster Labels       1st Most Common Venue  \
0      Emerson Hill               2  Construction & Landscaping   
1             Emery               2              Baseball Field   
2         Humberlea               2              Baseball Field   
3        Humber Bay               2  Construction & Landscaping   
4  King's Mill Park               2  Construction & Landscaping   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Farmers Market   Egyptian Restaurant     Electronics Store   
1         Women's Store        Farmers Market     Electronics Store   
2         Women's Store        Farmers Market     Electronics Store   
3        Baseball Field  Fast Food Restaurant   Empanada Restaurant   
4        Baseball Field  Fast Food Restaurant   Empanada Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
1   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
2   Empanada Restaurant    English Restaurant  Ethiopian Restaurant   
3    English Restaurant  Ethiopian Restaurant         Event Service   
4    English Restaurant  Ethiopian Restaurant         Event Service   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
0         Event Service           Event Space                Exhibit   
1         Event Service           Event Space                Exhibit   
2         Event Service           Event Space                Exhibit   
3           Event Space               Exhibit             Eye Doctor   
4           Event Space               Exhibit             Eye Doctor   

         Borough   Latitude  Longitude           City  
0  Staten Island  40.606794 -74.097762  New York City  
1     North York  43.724766 -79.532242        Totonto  
2     North York  43.724766 -79.532242        Totonto  
3      Etobicoke  43.636258 -79.498509        Totonto  
4      Etobicoke  43.636258 -79.498509        Totonto

<H1>Drop duplicates in clusters</H1>

In [48]:


cluster1_cities.drop_duplicates(inplace=True)
cluster2_cities.drop_duplicates(inplace=True)
cluster3_cities.drop_duplicates(inplace=True)
cluster4_cities.drop_duplicates(inplace=True)
cluster5_cities.drop_duplicates(inplace=True)



In [49]:
clust1_NYC     = cluster1_cities[cluster1_cities['City'] =='New York City']
clust1_Toronto = cluster1_cities[cluster1_cities['City'] =='Toronto']
len(clust1_NYC)

5

In [50]:


clust1_NYC= cluster1_cities[cluster1_cities['City']=='New York City']
clust1_Toronto = cluster1_cities[cluster1_cities['City']=='Toronto']

clust2_NYC= cluster2_cities[cluster2_cities['City']=='New York City']
clust2_Toronto = cluster2_cities[cluster2_cities['City']=='Toronto']

clust3_NYC= cluster3_cities[cluster3_cities['City']=='New York City']
clust3_Toronto = cluster3_cities[cluster3_cities['City']=='Toronto']

clust4_NYC= cluster4_cities[cluster4_cities['City']=='New York City']
clust4_Toronto = cluster4_cities[cluster4_cities['City']=='Toronto']


clust5_NYC= cluster5_cities[cluster5_cities['City']=='New York City']
clust5_Toronto = cluster5_cities[cluster5_cities['City']=='Toronto']

<h2>Let's a look at the contribution of Toronto and New York to each cluster.</h2>
<h2>Both cities contribute to clusters 1, 2,3 and 5.</h2>
<h2>Cluster 4 is 100% Toronto neighborhoods</h2>

In [51]:
print("Cluster 1:  New York members={},  Toronto members={}".format(len(clust1_NYC),(len(cluster1_cities)-len(clust1_NYC))))
print("Cluster 2:  New York members={},  Toronto members={}".format(len(clust2_NYC),(len(cluster2_cities)-len(clust2_NYC))))
print("Cluster 3:  New York members={},  Toronto members={}".format(len(clust3_NYC),(len(cluster3_cities)-len(clust3_NYC))))
print("Cluster 4:  New York members={},  Toronto members={}".format(len(clust4_NYC),(len(cluster4_cities)-len(clust4_NYC))))
print("Cluster 5:  New York members={},  Toronto members={}".format(len(clust5_NYC),(len(cluster5_cities)-len(clust5_NYC))))

Cluster 1:  New York members=5,  Toronto members=24
Cluster 2:  New York members=51,  Toronto members=41
Cluster 3:  New York members=1,  Toronto members=10
Cluster 4:  New York members=0,  Toronto members=5
Cluster 5:  New York members=249,  Toronto members=127


In [52]:
# all the venues have the same column names
cols=list(cluster_2.columns)
col = cols[2:12]

# extract venue names for each cluster
def venue_list(thisClust):
    LL=[]
    for c in col:
        L2 = list((thisClust[c]))
        LL.append(L2)

    LLL = list()
    for i in range(10):
        for j in LL[:][i]:
            LLL.append(j)
    return LLL

# a list of venues for each of the five clusters
# these lists will be used later to create new lists of venu counts (for each cluster)
Ven1 = venue_list(cluster_1)
Ven2 = venue_list(cluster_2)
Ven3 = venue_list(cluster_3)
Ven4 = venue_list(cluster_4)
Ven5 = venue_list(cluster_5)


## Let's try to get a feel of each cluster's '1st Most Common Venues'

<h2>'1st Most Common Venues in Cluster # 1'    [notice no restaurants in the '1st Most Common Venue']</h2>

In [53]:
c1 = cluster1_cities[['City','Neighborhood','1st Most Common Venue']]
v1=c1[['City','1st Most Common Venue']]
most1= pd.DataFrame(v1['1st Most Common Venue'])
most1['1st Most Common Venue'].value_counts(dropna=False)

Park                          17
Construction & Landscaping     3
Smoke Shop                     3
Airport                        2
Trail                          2
Bus Line                       1
Fast Food Restaurant           1
Name: 1st Most Common Venue, dtype: int64

<h2>'1st Most Common Venue' in Cluster # 2  [notice MANY restaurants in the '1st Most Common Venue']</h2>

In [54]:
c2 = cluster2_cities[['City','Neighborhood','1st Most Common Venue']]
v2=c2[['City','1st Most Common Venue']]
most2= pd.DataFrame(v2['1st Most Common Venue'])
most2['1st Most Common Venue'].value_counts(dropna=False)

Pizza Place                        35
Pharmacy                           11
Coffee Shop                        10
Grocery Store                       9
Café                                3
Caribbean Restaurant                3
Indian Restaurant                   2
Deli / Bodega                       2
Chinese Restaurant                  1
Spanish Restaurant                  1
Southern / Soul Food Restaurant     1
Empanada Restaurant                 1
Intersection                        1
Bank                                1
Bakery                              1
Fast Food Restaurant                1
Other Nightlife                     1
Playground                          1
Pub                                 1
Sandwich Place                      1
Baseball Field                      1
Cosmetics Shop                      1
Discount Store                      1
Rental Car Location                 1
Italian Restaurant                  1
Name: 1st Most Common Venue, dtype: int64

<h2>'1st Most Common Venue' in Cluster # 3   [notice no restaurants in the '1st Most Common Venue']'</h2>

In [55]:
c3 = cluster3_cities[['City','Neighborhood','1st Most Common Venue']]
v3=c3[['City','1st Most Common Venue']]
most3= pd.DataFrame(v3['1st Most Common Venue'])
most3['1st Most Common Venue'].value_counts(dropna=False)

Construction & Landscaping    9
Baseball Field                2
Name: 1st Most Common Venue, dtype: int64

<h2>'1st Most Common Venue' in Cluster # 4   [notice no restaurants in the '1st Most Common Venue']</h2>

In [56]:
c4 = cluster4_cities[['City','Neighborhood','1st Most Common Venue']]
v4=c4[['City','1st Most Common Venue']]
most4= pd.DataFrame(v4['1st Most Common Venue'])
most4['1st Most Common Venue'].value_counts(dropna=False)

Bank    5
Name: 1st Most Common Venue, dtype: int64

<h2>'1st Most Common Venue' in Cluster # 5   [notice there are some restaurants in the '1st Most Common Venue']</h2>

In [57]:
c5 = cluster5_cities[['City','Neighborhood','1st Most Common Venue']]
v5=c5[['City','1st Most Common Venue']].drop_duplicates()
most5= pd.DataFrame(v5['1st Most Common Venue'])
most5['1st Most Common Venue'].value_counts(dropna=False)

Rental Car Location            2
Convenience Store              2
Bakery                         2
Clothing Store                 2
Middle Eastern Restaurant      2
Cosmetics Shop                 2
Bus Station                    2
Liquor Store                   2
Fried Chicken Joint            2
Chinese Restaurant             2
Café                           2
Furniture / Home Store         2
Dessert Shop                   2
Indian Restaurant              2
Greek Restaurant               2
Coffee Shop                    2
Pizza Place                    2
Fast Food Restaurant           2
Grocery Store                  2
Bar                            2
Women's Store                  2
Intersection                   2
Caribbean Restaurant           2
Food Truck                     2
Ramen Restaurant               1
Burger Joint                   1
Gym                            1
Trail                          1
Athletics & Sports             1
Park                           1
Peruvian R

In [58]:
L=cluster_2[['1st Most Common Venue',
 '2nd Most Common Venue',
 '3rd Most Common Venue',
 '4th Most Common Venue',
 '5th Most Common Venue',
 '6th Most Common Venue',
 '7th Most Common Venue',
 '8th Most Common Venue',
 '9th Most Common Venue',
 '10th Most Common Venue']]

In [59]:
# define the dataframe columns

col_names1 = ['venues_Cluster_1'] 
col_names2 = ['venues_Cluster_2'] 
col_names3 = ['venues_Cluster_3'] 
col_names4 = ['venues_Cluster_4'] 
col_names5 = ['venues_Cluster_5'] 

# instantiate the dataframes for each venue 

clust1_venues = pd.DataFrame(Ven1,columns=col_names1)
clust2_venues = pd.DataFrame(Ven2,columns=col_names2)
clust3_venues = pd.DataFrame(Ven3,columns=col_names3)
clust4_venues = pd.DataFrame(Ven4,columns=col_names4)
clust5_venues = pd.DataFrame(Ven5,columns=col_names5)

# count each venue category of each cluster
c1_venue_count =pd.DataFrame(clust1_venues['venues_Cluster_1'].value_counts(dropna=False))
c2_venue_count =pd.DataFrame(clust2_venues['venues_Cluster_2'].value_counts(dropna=False))
c3_venue_count =pd.DataFrame(clust3_venues['venues_Cluster_3'].value_counts(dropna=False))
c4_venue_count =pd.DataFrame(clust4_venues['venues_Cluster_4'].value_counts(dropna=False))
c5_venue_count =pd.DataFrame(clust5_venues['venues_Cluster_5'].value_counts(dropna=False))

<h1>Let's refine cluster characterization by expanding venue count to include all top 10 venues.</h1>

<h3>---------------------------------------</h3>

<h3>The BAKERY LOCATION SEARCH HYPOTHEISIS</h3>

<h3>The hypothesis is:  Venues that neighborhoods that have many restaurants are food friendly' neighborhoods and </h3>
<h3>'food friendly' neighborhoods that currently having no bakeries may be potentially a good location for a new bakery.</h3>
<h3>Let's look at each cluster's restaurant venues counts.</h3>
<h3>----------------------------------------------------------------------------------------------------------------</h3>

<h3>Is Cluster 1 a 'restaurant friendly' cluster ?  YES</h3>
<h3>There are 100 restaurants and 3 bakeries. Approximatetly a ratio 33:1  restaurants to bakeries  </h3>

In [60]:
c1_rest= pd.DataFrame(clust1_venues[clust1_venues['venues_Cluster_1'].str.contains('Restaurant')])
Clust_1_rest_types =pd.DataFrame(c1_rest['venues_Cluster_1'].value_counts(dropna=False))
c1_bakery= pd.DataFrame(clust1_venues[clust1_venues['venues_Cluster_1'].str.contains('Bakery')])
Clust_1_bakery_types =pd.DataFrame(c1_bakery['venues_Cluster_1'].value_counts(dropna=False))
print(Clust_1_rest_types)
print('---------------------------------------')

print('There are {} restaurants and {} bakeries.'.format(len(c1_rest),len(c1_bakery)))


                             venues_Cluster_1
Empanada Restaurant                        28
English Restaurant                         28
Ethiopian Restaurant                       26
Falafel Restaurant                          6
Egyptian Restaurant                         6
Fast Food Restaurant                        2
Sushi Restaurant                            2
South American Restaurant                   1
Eastern European Restaurant                 1
---------------------------------------
There are 100 restaurants and 3 bakeries.


<h3>Is Cluster 2 a 'restaurant friendly' cluster ?   YES</h3>
<h3>There are 191 restaurants and 14 bakeries. The ratio of restaurant to bakery = 14:1 </h3>

In [61]:
c2_rest= pd.DataFrame(clust2_venues[clust2_venues['venues_Cluster_2'].str.contains('Restaurant')])
Clust_2_rest_types =pd.DataFrame(c2_rest['venues_Cluster_2'].value_counts(dropna=False))
c2_bakery= pd.DataFrame(clust2_venues[clust2_venues['venues_Cluster_2'].str.contains('Bakery')])
Clust_2_bakery_types =pd.DataFrame(c2_bakery['venues_Cluster_2'].value_counts(dropna=False))
print(Clust_2_rest_types)
print('---------------------------------------')
print('There are {} restaurants and {} bakeries.'.format(len(c2_rest),len(c2_bakery)))

                                 venues_Cluster_2
Fast Food Restaurant                           26
Chinese Restaurant                             23
Empanada Restaurant                            19
English Restaurant                             19
Restaurant                                     13
Italian Restaurant                             11
Mexican Restaurant                              9
Ethiopian Restaurant                            8
Egyptian Restaurant                             7
Caribbean Restaurant                            7
American Restaurant                             6
Sushi Restaurant                                5
Middle Eastern Restaurant                       5
Latin American Restaurant                       5
Indian Restaurant                               4
Thai Restaurant                                 4
Falafel Restaurant                              4
Asian Restaurant                                3
Spanish Restaurant                              3


<h3>Cluster 3 is restaurant friendly?  YES</h3>
<h3>There are 42 restaurants and 0 bakeries   </h3>

In [62]:
c3_rest= pd.DataFrame(clust3_venues[clust3_venues['venues_Cluster_3'].str.contains('Restaurant')])
Clust_3_rest_types =pd.DataFrame(c3_rest['venues_Cluster_3'].value_counts(dropna=False))
c3_bakery= pd.DataFrame(clust3_venues[clust3_venues['venues_Cluster_3'].str.contains('Bakery')])
Clust_3_bakery_types =pd.DataFrame(c3_bakery['venues_Cluster_3'].value_counts(dropna=False))
print(Clust_3_rest_types)
print('---------------------------------------')
#print(Clust_1_bakery_types)
#print('---------------------------------------')
print('There are {} restaurants and {} bakeries.'.format(len(c3_rest),len(c3_bakery)))



                      venues_Cluster_3
Empanada Restaurant                 11
English Restaurant                  11
Ethiopian Restaurant                11
Fast Food Restaurant                 8
Egyptian Restaurant                  1
---------------------------------------
There are 42 restaurants and 0 bakeries.


<h3>Is Cluster 4 'restaurant friendly' cluster ?</h3>
<h3>There are 15 restaurants, but there no Bakeries. </h3>

In [63]:
c4_rest= pd.DataFrame(clust4_venues[clust4_venues['venues_Cluster_4'].str.contains('Restaurant')])
Clust_4_rest_types =pd.DataFrame(c4_rest['venues_Cluster_4'].value_counts(dropna=False))
c4_bakery= pd.DataFrame(clust4_venues[clust4_venues['venues_Cluster_4'].str.contains('Bakery')])
Clust_4_bakery_types =pd.DataFrame(c4_bakery['venues_Cluster_4'].value_counts(dropna=False))
print(Clust_4_rest_types)
print('---------------------------------------')
print('There are {} restaurants and {} bakeries.'.format(len(c4_rest),len(c4_bakery)))

                      venues_Cluster_4
English Restaurant                   5
Empanada Restaurant                  5
Ethiopian Restaurant                 5
---------------------------------------
There are 15 restaurants and 0 bakeries.


<h3>Is Cluster 5 restaurant friendly cluster ?   YES,  but there any many bakeries </h3>

<h3>There are 1005 restaurants and 101 bakeries.   Approximatetly a ratio 10:1  restaurants to bakeries</h3>

In [64]:
c5_rest= pd.DataFrame(clust5_venues[clust5_venues['venues_Cluster_5'].str.contains('Restaurant')])
Clust_5_rest_types =pd.DataFrame(c5_rest['venues_Cluster_5'].value_counts(dropna=False))
c5_bakery= pd.DataFrame(clust5_venues[clust5_venues['venues_Cluster_5'].str.contains('Bakery')])
Clust_5_bakery_types =pd.DataFrame(c5_bakery['venues_Cluster_5'].value_counts(dropna=False))
print(Clust_5_rest_types)
print('---------------------------------------')
print('There are {} restaurants and {} bakeries.'.format(len(c5_rest),len(c5_bakery)))

                                 venues_Cluster_5
Italian Restaurant                            100
Chinese Restaurant                             82
Ethiopian Restaurant                           67
English Restaurant                             67
Empanada Restaurant                            62
American Restaurant                            61
Fast Food Restaurant                           60
Mexican Restaurant                             57
Restaurant                                     47
Sushi Restaurant                               37
Caribbean Restaurant                           30
Japanese Restaurant                            30
Thai Restaurant                                23
Seafood Restaurant                             23
Latin American Restaurant                      20
Vietnamese Restaurant                          17
Asian Restaurant                               17
Indian Restaurant                              17
French Restaurant                              16


<h3>Cluster 1, 3 and 4 would be worth evaluating initially.</h3>
<h3>Cluster 1 has a high restaurant to bakery ratio of 33.  
<h3>Clusters 3 and 4 have many restauants and no bakeries.</h3>
<h3>Which cities/neighborhoods are in Clusters 1, 3 and 4  ?</h3>

In [65]:
city_neigh_134 = pd.concat([cluster1_cities[['City','Neighborhood']],
                           cluster3_cities[['City','Neighborhood']],
                           cluster4_cities[['City','Neighborhood']]],axis=0)

city_neigh_134.reset_index(drop=True)

City              Neighborhood
0   New York City              Clason Point
1   New York City                Somerville
2   New York City                 Todt Hill
3   New York City             Randall Manor
4   New York City                 Bayswater
5         Totonto                 Parkwoods
6         Totonto       Caledonia-Fairbanks
7         Totonto              East Toronto
8         Totonto               CFB Toronto
9         Totonto            Downsview East
10        Totonto                 Downsview
11        Totonto                North Park
12        Totonto               Upwood Park
13        Totonto             Lawrence Park
14        Totonto                    Weston
15        Totonto           York Mills West
16        Totonto         Forest Hill North
17        Totonto          Forest Hill West
18        Totonto         Kingsview Village
19        Totonto      Martin Grove Gardens
20        Totonto          Richview Gardens
21        Totonto              St. Phillips
22        Totonto           Agincourt North
23        Totonto           L'Amoreaux East
24        Totonto                  Milliken
25        Totonto              Steeles East
26        Totonto              The Kingsway
27        Totonto           Montgomery Road
28        Totonto            Old Mill North
29  New York City              Emerson Hill
30        Totonto                     Emery
31        Totonto                 Humberlea
32        Totonto                Humber Bay
33        Totonto          King's Mill Park
34        Totonto  Kingsway Park South East
35        Totonto                 Mimico NE
36        Totonto            Old Mill South
37        Totonto        The Queensway East
38        Totonto     Royal York South East
39        Totonto                  Sunnylea
40        Totonto                Cloverdale
41        Totonto                 Islington
42        Totonto              Martin Grove
43        Totonto          Princess Gardens
44        Totonto           West Deane Park

<h4>Conclusions:</h4>
<h4>Toronto and New York City both have urban neighborhoods.  During the clustering creation a fixed radius (300 meters) was used.</h4>
<h4>The neighborhoods in Clusters 1, 3 and 4 are 'restaurant friendly' and currently have either no bakeries or have a high</h4> 
<h4>restaurant to bakery ratio and represent a potential favorable location for a new bakery.</h4>
<h4>There are 517 neighborhoods in Toronto and New York City.</h4>
<h4>Using the Kmeans algorithm, the number of neighborhood in these two cities that</h4>
<h4>is 44.  So, reducing potential neighborhoods for a new bakery location from 517 to 44 using machine learning.</h4>
<h4>This is just a 'quick look'.  Future studies would include demographic data.</h4>